File: src/embedder.rs
use ort::session::Session;
use ort::session::builder::GraphOptimizationLevel as OptLevel;
use ort::value::Tensor;
use tokenizers::Tokenizer;

pub struct LocalEmbedder {
    session: Session,
    tokenizer: Tokenizer,
}

impl LocalEmbedder {
    pub fn new() -> Result<Self, String> {
        let home_dir = dirs::home_dir().ok_or("Could not resolve OS home directory")?;
        let model_dir = home_dir.join(".celial").join("models");

        let model_path = model_dir.join("model.onnx");
        let tokenizer_path = model_dir.join("tokenizer.json");

        if !model_path.exists() || !tokenizer_path.exists() {
            return Err(format!("Embedding model missing. Expected at {:?}", model_dir));
        }

        let tokenizer = Tokenizer::from_file(tokenizer_path)
            .map_err(|e| format!("Tokenizer error: {}", e))?;

        let session = Session::builder()
            .map_err(|e| e.to_string())?
            .with_optimization_level(OptLevel::Level3)
            .map_err(|e| e.to_string())?
            .commit_from_file(model_path)
            .map_err(|e| e.to_string())?;

        Ok(Self {
            session,
            tokenizer,
        })
    }

    pub fn embed(&mut self, text: &str) -> Result<Vec<f32>, String> {
        let mut tokenizer = self.tokenizer.clone();
        tokenizer.with_truncation(Some(tokenizers::TruncationParams {
            max_length: 512,
            ..Default::default()
        })).map_err(|e| format!("Truncation config failed: {}", e))?;

        let encoding = tokenizer.encode(text, true)
            .map_err(|e| format!("Encoding failed: {}", e))?;

        let input_ids = encoding
            .get_ids()
            .iter()
            .map(|&x| x as i64)
            .collect::<Vec<_>>();
        let attention_mask = encoding
            .get_attention_mask()
            .iter()
            .map(|&x| x as i64)
            .collect::<Vec<_>>();

        let seq_len = input_ids.len();

        let input_ids_tensor = Tensor::from_array(([1usize, seq_len], input_ids))
            .map_err(|e| e.to_string())?;
        let mask_tensor = Tensor::from_array(([1usize, seq_len], attention_mask))
            .map_err(|e| e.to_string())?;

        let outputs = self
            .session
            .run(ort::inputs! {
                "input_ids" => input_ids_tensor,
                "attention_mask" => mask_tensor,
            })
            .map_err(|e| e.to_string())?;

        let (_, embeddings_data) = outputs[0]
            .try_extract_tensor::<f32>()
            .map_err(|e| e.to_string())?;

        let expected = 1usize * seq_len * 768;
        if embeddings_data.len() < expected {
            return Err("Output tensor shorter than expected".into());
        }

        let mut pooled = vec![0.0f32; 768];
        for i in 0..seq_len {
            let offset = i * 768;
            for d in 0..768 {
                pooled[d] += embeddings_data[offset + d];
            }
        }
        for d in 0..768 {
            pooled[d] /= seq_len as f32;
        }

        let norm: f32 = pooled.iter().map(|x| x * x).sum::<f32>().sqrt();
        if norm > 0.0 {
            for d in 0..768 {
                pooled[d] /= norm;
            }
        }

        Ok(pooled)
    }
}

File: src/embedder.py
from pathlib import Path
import numpy as np
import onnxruntime as ort
from tokenizers import Tokenizer


def bootstrap_embedding_model() -> tuple[Path, Path]:
    """Resolves local storage bounds and retrieves the ONNX embedding model assets if needed."""
    import urllib.request
    home_dir = Path.home()
    model_dir = home_dir / ".celial" / "models"

    model_path = model_dir / "model.onnx"
    tokenizer_path = model_dir / "tokenizer.json"

    if not model_path.exists() or not tokenizer_path.exists():
        model_dir.mkdir(parents=True, exist_ok=True)
        MODEL_URL = "https://huggingface.co/Qdrant/all-MiniLM-L6-v2-onnx/resolve/main/model.onnx"
        TOKENIZER_URL = "https://huggingface.co/Qdrant/all-MiniLM-L6-v2-onnx/resolve/main/tokenizer.json"
        
        try:
            if not model_path.exists():
                print(f"Downloading ONNX model to {model_path} (one-time operation)...")
                urllib.request.urlretrieve(MODEL_URL, model_path)
            if not tokenizer_path.exists():
                print(f"Downloading tokenizer to {tokenizer_path}...")
                urllib.request.urlretrieve(TOKENIZER_URL, tokenizer_path)
        except Exception as e:
            raise RuntimeError(f"Failed to download embedding model: {e}")
            
    return model_path, tokenizer_path


class LocalEmbedder:

    def __init__(self):
        """
        Cold-boots the embedding environment by initializing the ONNX optimization pipeline.
        """
        model_path, tokenizer_path = bootstrap_embedding_model()

        try:
            self.tokenizer = Tokenizer.from_file(str(tokenizer_path))
        except Exception as e:
            raise RuntimeError(f"Tokenizer error: {e}")

        # Replicating Level 3 graph optimization configuration
        opts = ort.SessionOptions()
        opts.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_ALL

        try:
            self.session = ort.InferenceSession(str(model_path),
                                                sess_options=opts)
        except Exception as e:
            raise RuntimeError(f"ONNX Session compilation failed: {e}")


    def embed(self, text: str) -> list[float]:
        """
        Encodes source text, executes tensor graph evaluation, 
        and applies vectorized mean-pooling with L2 normalization constraints.
        """
        # Constrain context window boundary inline
        self.tokenizer.enable_truncation(max_length=512)

        try:
            encoding = self.tokenizer.encode(text)
        except Exception as e:
            raise RuntimeError(f"Encoding failed: {e}")

        # Construct 2D tensor matrices [1, sequence_length] directly
        input_ids = np.array([encoding.ids], dtype=np.int64)
        attention_mask = np.array([encoding.attention_mask], dtype=np.int64)

        try:
            outputs = self.session.run(None, {
                "input_ids": input_ids,
                "attention_mask": attention_mask
            })
        except Exception as e:
            raise RuntimeError(f"ONNX inference execution failed: {e}")

        # Outputs shape expected: (1, seq_len, 768)
        embeddings_data = outputs[0]

        if embeddings_data.ndim != 3 or embeddings_data.shape[2] not in (384, 768):
            raise ValueError(
                "Output tensor shape deviates from the expected 384 or 768-dimensional space."
            )

        # Vectorized Mean Pooling along the sequence dimension axis (axis 1)
        pooled = np.mean(embeddings_data, axis=1)[0]

        # Vectorized L2 Normalization constraint satisfaction
        norm = np.linalg.norm(pooled)
        if norm > 0.0:
            pooled = pooled / norm

        return pooled.tolist()
